hub/core/transform/transform.py

```diff
@@ -0,0 +1,164 @@
+        pipeline (Sequence[Callable]): A Sequence of functions to apply to each element of data_in to generate output dataset.
+            The output of each function should either be a dictionary or a list/tuple of dictionaries.
+            The last function has added restriction that keys in the output of each sample should be same as the tensors present in the ds_out object.
+        ds_out (Dataset): The dataset object to which the transform will get written.
+            Should have all keys being generated in output already present as tensors. It's initial state should be either:-
+            - Empty i.e. all tensors have no samples. In this case all samples are added to the dataset.
+            - All tensors are populated and have sampe length. In this case new samples are appended to the dataset.
+        pipeline_kwargs (List[dict], optional): A list containing extra arguments to be passed to the pipeline functions.
+            If more kwargs than functions in pipeline, extra kwargs are ignored, if less kwargs, they are matched to only the starting functions.
+            To use on non-continuous functions fill empty dict. Eg. pipeline=[fn1,fn2,fn3], kwargs=[{"a":5},{},{"c":1,"s":7}], only applies to fn1 and fn3.
+        scheduler (str): The scheduler to be used to compute the transformation. Currently can be one of 'threaded' and 'processed'.
+        workers (int): The number of workers to use for performing the transform. Defaults to 1.
+
+    Raises:
+        InvalidInputDataError: If ds_in passed to transform is invalid. It should support __getitem__ and __len__ operations.
+        TensorMismatchError: If one or more of the outputs generated during transform contain different tensors than the ones present in the output 'ds_out' provided to transform.
+        UnsupportedSchedulerError: If the scheduler passed is not recognized.
+        InvalidTransformOutputError: If the output of any step in a transformation isn't dictionary or a list/tuple of dictionaries.
+    """
+    ds_out.flush()
+
+    if not hasattr(data_in, "__getitem__"):
+        raise InvalidInputDataError("__getitem__")
+    if not hasattr(data_in, "__len__"):
+        raise InvalidInputDataError("__len__")
+
+    # TODO: this check doesn't work currently. Will work once AL-1092 is merged and can be uncommented.
+    # for tensor in tensors:
+    #     assert len(ds_out[tensor]) == len(ds_out)
+
+    workers = max(workers, 1)
+
+    if scheduler == "threaded":
+        compute: ComputeProvider = ThreadProvider(workers)
+    elif scheduler == "processed":
+        compute = ProcessProvider(workers)
+    else:
+        raise UnsupportedSchedulerError(scheduler)
+
+    base_storage = get_base_storage(ds_out.storage)
+    tensors = set(ds_out.meta.tensors)
+
+    pipeline_kwargs = equalize_pipeline_kwargs(pipeline_kwargs, pipeline)
```

Comment:
Equalize? I have no idea what this does. I think we can come up with a better name